### DISCUSSION TITLE
Error in watchdog loop

### DISCUSSION THREAD

**Discussion by TommyTomson585** [FreeOpcUa/opcua-asyncio#1494] [Q&A]
Hey there,

I am using the library to connect to a OPC Server of a welding machine. The  client creates a subcription and resubscribes if it looses the connection.

I am using the asyncua in the version 1.0.3 an python 3.11

Let me say  that I am just a "user" of the code. My knowledge is very basic or low level.

Here a snip of the code. It is just the "handler class" and the definition of the opcua_client() method.

`
```
 import asyncio
from asyncua import Client, Server, ua, Node
import datetime
from time import sleep
import logging

....

class SubHandler:
    def __init__(self):
        self.initial = True

        

        
     
          
   
async def datachange_notification(self, node: Node, val, data):
       
            global valCounter1, valCounter2
        
            async with datachange_notification_queue_lock:
                datachange_notification_queue.append((node,val,data))
        
                print("Node hat den Status geaendert. Aktueller Wert:",node, val)

                if node == itemlist[0]:
                    
                    if val is not None:
                        
                        if valCounter1 is None:
                            
                            valCounter1= val
                        
                        else:
                            
                            if val > valCounter1:
            
                                valCounter1=val
            
                   
                                await asyncio.sleep(1)
                
                                if station1.getValue() == None:
                                    print("Station1 leer" )
                                else:   
                                    print("Station1"  + station1.getValue())

                                if station2.getValue() == None:
                                    print("Station2 leer")
                                else:    
    
                                    print("Station2"  + station2.getValue())
               
                                if station1.getValue() == station2.getValue():
                                    counterBerghaus.doInkre()
                
                                elif station1.getValue() != station2.getValue():
                                    counterBerghaus.inkre()
                        
                    #if self.initial==True:
                        
                        #self.initial=False
                        
                    #else:
                                
                                await countState.set_value(counterBerghaus.getValue())
                  
                                await write()
                                await aufzeichner3()
                
                
                elif node == itemlist[1]:
                    
                    if val is not None:
                        
                        if valCounter2 is None:
                            
                            valCounter2= val
                            
                        else:
                            
                            if val > valCounter2:
                      
                
                                valCounter2 = val
                                await asyncio.sleep(1)
                
                elif node == itemlist[2]:
               
                    station1.setValue(val)
                    await aufzeichner3()
                
                elif node == itemlist[3]:
                
                    station2.setValue(val)
                    await aufzeichner3()
            
                elif node == itemlist[4]:
                    valState=val
                    valState=bool(valState)
                    global statusRob
                    statusRob = valState
                    
                    
                    await servState.set_value(valState)
                    await aufzeichner3()
                
        
            
      
    def event_notification(self, event):
        
        print("Python: New event", event)

async def opcua_client():
    
       
    
        client = Client(url=ServerUrl)
        
        
        
        itemlist.append(client.get_node(counter1))
        itemlist.append(client.get_node(counter2))
        itemlist.append(client.get_node(product1))
        itemlist.append(client.get_node(product2))
        itemlist.append(client.get_node(status))
        #itemlist.append(client.get_node(serverStatus))
       
       
       
       
        #client=Client("opc.tcp://Berghaus-WS26:4980/Softing_dataFEED_OPC_Suite_Configuration1")
        
        handler= SubHandler()
        subscription = None
        
        case = 0
        subscription_handle_list=[]
        idx = 0
        
        while 1:
            if case ==1:
                #connect
                print("connecting...")
                try:
                    await client.connect()
                    print("connected!")
                    case = 2
                    
                except:
                    print("connection error!")
                    case = 1
                    await  asyncio.sleep(2)
            elif case == 2:
                
                #subscribing nodes and events
                
                print("subscribing nodes and events.....")
                
                try:
                    subscription = await client.create_subscription(200, handler)
                    subscription_handle_list = []
                    if nodes_to_subscribe:
                        for node in nodes_to_subscribe:
                            handle = await subscription.subscribe_data_change(client.get_node(node))
                            subscription_handle_list.append(handle)
                    print("subscribed!!")
                    
                    case = 3
                except:
                    print("subscription error!!")
                    
                    case= 4
                    await asyncio.sleep(0)
            elif case == 3:
                
                try:
                    #service_level = await client.get_node("ns=12;s=Robo3_OPC.nsuri=http://opcfoundation.org/UA/;n=2259").get_value()
                    service_level = await client.get_node("ns=0;i=2256").get_value()
                    
                    
                    if service_level:
                        
                        case = 3
                        
                    else:
                        
                        case = 4
                    await asyncio.sleep(2)
                    
                except:
                    
                    case = 4
            
            elif case == 4:
                
                print("unsubscribing....!!")
                
                try:
                    
                    if subscription_handle_list:
                        for handle in subscription_handle_list:
                            await subscription.unsubscribe(handle)
                    
                    await subscription.delete()
                    print("unsubscribed!!!")
                
                except:
                    
                    print("unsubscription error!!!")
                    
                    subscription = None
                    subscription_handle_list = []
                    await asyncio.sleep(0)
                
                print("disconnecting...")
                
                try:
                    
                    await client.disconnect()
                
                except:
                    
                    print ("disconnection error!")
                
                case = 0
                
            else:
                
                case = 1
                await asyncio.sleep(2)
```

`

The client is working fine but from time to time I have an exception:

Here a part of the logfile when the error appears:

.
.

> .
> count_inkretwo
> 14094
> Node hat den Status geaendert. Aktueller Wert: ns=5;s=MotionDeviceSystem.Controllers.Controller_1.ProcessData.Partcounter.Partcounter_St2 28703
> Node hat den Status geaendert. Aktueller Wert: ns=5;s=MotionDeviceSystem.Controllers.Controller_1.ProcessData.State.ARC_ON 0
> Node hat den Status geaendert. Aktueller Wert: ns=5;s=MotionDeviceSystem.Controllers.Controller_1.ProcessData.Productinfo.Productinfo_St1 RG15844524
> Node hat den Status geaendert. Aktueller Wert: ns=5;s=MotionDeviceSystem.Controllers.Controller_1.ProcessData.Productinfo.Productinfo_St2 RG-15844524
> Node hat den Status geaendert. Aktueller Wert: ns=5;s=MotionDeviceSystem.Controllers.Controller_1.ProcessData.State.ARC_ON 1
> Node hat den Status geaendert. Aktueller Wert: ns=5;s=MotionDeviceSystem.Controllers.Controller_1.ProcessData.State.ARC_ON 0
> 
> Error in watchdog loop
> Traceback (most recent call last):
>   File "C:\Users\Thomas.Hartz\AppData\Local\Programs\Python\Python311\Lib\asyncio\tasks.py", line 490, in wait_for
>     return fut.result()
>            ^^^^^^^^^^^^
> asyncio.exceptions.CancelledError
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "C:\Users\Thomas.Hartz\AppData\Local\Programs\Python\Python311\Lib\site-packages\asyncua\client\client.py", line 520, in _monitor_server_loop
>     _ = await self.nodes.server_state.read_value()
>         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\Thomas.Hartz\AppData\Local\Programs\Python\Python311\Lib\site-packages\asyncua\common\node.py", line 179, in read_value
>     result = await self.read_data_value()
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\Thomas.Hartz\AppData\Local\Programs\Python\Python311\Lib\site-packages\asyncua\common\node.py", line 190, in read_data_value
>     return await self.read_attribute(ua.AttributeIds.Value, None, raise_on_bad_status)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\Thomas.Hartz\AppData\Local\Programs\Python\Python311\Lib\site-packages\asyncua\common\node.py", line 304, in read_attribute
>     result = await self.session.read(params)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\Thomas.Hartz\AppData\Local\Programs\Python\Python311\Lib\site-packages\asyncua\client\ua_client.py", line 397, in read
>     data = await self.protocol.send_request(request)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\Thomas.Hartz\AppData\Local\Programs\Python\Python311\Lib\site-packages\asyncua\client\ua_client.py", line 160, in send_request
>     data = await asyncio.wait_for(self._send_request(request, timeout, message_type), timeout if timeout else None)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\Thomas.Hartz\AppData\Local\Programs\Python\Python311\Lib\asyncio\tasks.py", line 492, in wait_for
>     raise exceptions.TimeoutError() from exc
> TimeoutError
> DataChange subscription has no status_change_notification method
> unsubscribing....!!
> unsubscription error!!!
> disconnecting...
> disconnection error!
> connecting...
> connection error!
> connecting...
> connection error!
> connecting...
> connection error!
> connecting...
> connection error!
> connecting...
> connection error!
> connecting...
> connected!
> subscribing nodes and events.....
> Revised values returned differ from subscription values: CreateSubscriptionResult(SubscriptionId=1155306408, RevisedPublishingInterval=200.0, RevisedLifetimeCount=18000, RevisedMaxKeepAliveCount=6000)
> subscribed!!
> Node hat den Status geaendert. Aktueller Wert: ns=5;s=MotionDeviceSystem.Controllers.Controller_1.ProcessData.Partcounter.Partcounter_St1 28676
> Station1RG15844524
> Station2RG-15844524
> Count_inkre
> 14094
> Node hat den Status geaendert. Aktueller Wert: ns=5;s=MotionDeviceSystem.Controllers.Controller_1.ProcessData.Partcounter.Partcounter_St2 28703
> Node hat den Status geaendert. Aktueller Wert: ns=5;s=MotionDeviceSystem.Controllers.Controller_1.ProcessData.State.ARC_ON 1
> Node hat den Status geaendert. Aktueller Wert: ns=5;s=MotionDeviceSystem.Controllers.Controller_1.ProcessData.Productinfo.Productinfo_St1 RG15844524
> Node hat den Status geaendert. Aktueller Wert: ns=5;s=MotionDeviceSystem.Controllers.Controller_1.ProcessData.Productinfo.Productinfo_St2 RG-15844524

Has anyone an idea why this error in the watchdog loop appears? 

Thanks  ahead for your support.

Kind regards 

Thomas

**Comment by schroeder-** [comment]
The watchdog loop is to detect if the connection is lost. So this error is triggered if the server is not reachable.

**Reply by TommyTomson585** [reply]
Thank you for your reply.

Okay I got it. But shouldn't that be catched by an exception?

**Reply by schroeder-** [reply]
you can call check_connection periodical, this will throw the exception. Or add status_change_notification in your subscription handler.
